numpyro/mcmc.py

```diff
@@ -677,31 +678,42 @@ def __init__(self,
         if num_chains > 1 or "CI" in os.environ or "PYTEST_XDIST_WORKER" in os.environ:
             self.progress_bar = False
 
+        self._collect_fields = ('z',)
         self._samples = None
         self._samples_flat = None
 
-    def _single_chain_mcmc(self, init, args=(), kwargs={}):
+    def _single_chain_mcmc(self, init, collect_fields=('z',), collect_warmup=False, args=(), kwargs={}):
         rng, init_params = init
         hmc_state, constrain_fn = self.sampler.init(rng, self.num_warmup, init_params,
                                                     model_args=args, model_kwargs=kwargs)
         if self.constrain_fn is None:
             self.constrain_fn = identity if constrain_fn is None else constrain_fn
-        samples = fori_collect(self.num_warmup, self.num_warmup + self.num_samples,
+        collect_fn = attrgetter(*collect_fields)
+        lower = 0 if collect_warmup else self.num_warmup
+        samples = fori_collect(lower, self.num_warmup + self.num_samples,
                                self.sampler.sample,
                                hmc_state,
-                               transform=lambda x: self.constrain_fn(x.z),
+                               transform=collect_fn,
                                progbar=self.progress_bar,
                                progbar_desc=functools.partial(get_progbar_desc_str, self.num_warmup),
                                diagnostics_fn=get_diagnostics_str)
+        if len(collect_fields) == 1:
+            samples = (samples,)
+        samples = dict(zip(collect_fields, samples))
+        samples['z'] = self.constrain_fn(samples['z'])
         return samples
 
-    def run(self, rng, *args, init_params=None, **kwargs):
+    def run(self, rng, *args, collect_fields=('z',), collect_warmup=False, init_params=None, **kwargs):
         """
         Run the MCMC samplers and collect samples.
 
         :param random.PRNGKey rng: Random number generator key to be used for the sampling.
         :param args: Arguments to be provided to the :meth:`numpyro.mcmc.MCMCKernel.init` method.
             These are typically the arguments needed by the `model`.
+        :param tuple, list collect_fields: Fields from :data:`numpyro.mcmc.HMCState` to collect
+            during the MCMC run. By default, only the latent sample sites `z` is collected.
```

Comment:
nit: I think that this won't be rendered correctly with sphinx. the better way is
```
:param collect_fields:
:type collect_fields: tuple or list
```

Reply:
Done.